### PR TITLE
Rewrite better-git PHP bot in golang

### PR DESCRIPTION
**Pull request by fetzi**
add a complete rewrite of the better-git PHP bot

### CODE REVIEW

File: internal/pkg/handler/handler.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+	switch payload.(type) {
+	case gitlab.PushEventPayload:
+		event := payload.(gitlab.PushEventPayload)
+
+		filters := filter.GetFilters(handler.conf)
+
+		for _, filter := range filters {
+			if filter.Filter(event) {
+				w.WriteHeader(http.StatusOK)
```
Comment (Munsio) [inline internal/pkg/handler/handler.go]: shouldn't this be an "unprocessable entity" or something like that including a body so the trigger source would know that what he sent is not really processed

File: internal/pkg/handler/handler.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+				return
+			}
+		}
+
+		validators := validator.GetValidators()
+
+		for _, commit := range event.Commits {
+
+			if handler.refStore.Exists(commit.ID) {
```
Comment (Munsio) [inline internal/pkg/handler/handler.go]: shouldn't this also check on `event.Project.PathWithNamespace`, like in the PUT below

File: internal/pkg/handler/handler.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+		conf,
+		commitLogger,
+		notifier,
+		refStore,
+	}
+}
+
+// ServeHTTP handles the gitlab hook request
+func (handler GitlabHookHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	hook, _ := gitlab.New(gitlab.Options.Secret(handler.conf.GitlabToken))
```
Comment (Munsio) [inline internal/pkg/handler/handler.go]: Set the hook as property of the GitlabHookHandler. So you instantiate it once in the new function and not every time the hook gets triggered.

File: internal/pkg/handler/handler.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Printf("%+v", err)
+		return
+	}
+
+	switch payload.(type) {
+	case gitlab.PushEventPayload:
+		event := payload.(gitlab.PushEventPayload)
+
+		filters := filter.GetFilters(handler.conf)
```
Comment (Munsio) [inline internal/pkg/handler/handler.go]: filters and validators can also be defined as private properties of the Handler itself.

File: internal/pkg/handler/handler.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+					"project":   event.Project.PathWithNamespace,
+					"valid":     true,
+				}).Info("Successfully validated commit")
+			}
+
+			handler.refStore.Put(commit.ID, event.Project.PathWithNamespace)
+		}
+	}
+
+	w.WriteHeader(http.StatusOK)
```
Comment (Munsio) [inline internal/pkg/handler/handler.go]: If there is no body use 204-NoContent

File: internal/pkg/handler/handler.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+func respondSuccess(message string, w http.ResponseWriter) {
```
Comment (Munsio) [inline internal/pkg/handler/handler.go]: unused function?

File: internal/pkg/refstore/refstore.go [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package refstore
+
+// RefStore interface
+type RefStore interface {
+	Put(key, value string)
```
Comment (Munsio) [inline internal/pkg/refstore/refstore.go]: Okay - the chance that commit hashes collide is probably 0% but why storing a value when we don't use it

File: internal/pkg/validator/validator.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+	return errors.New("Body cannot be extracted")
+}
+
+// TicketNumberValidator defines a validator for ticket numbers in commits
+type TicketNumberValidator struct{}
+
+// Validate checks for ticket numbers in the commit message and for their position
+func (tnv TicketNumberValidator) Validate(c gitlab.Commit) error {
+	regex, _ := regexp.Compile("(.*)\n\n([\\S\\s]*)")
```
Comment (Munsio) [inline internal/pkg/validator/validator.go]: would it be possible to use a seperate function which always returns subject and body?
There is much duplicated code

File: internal/app/commitbot/commitbot.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	notifier := notifier.NewSlackNotifier(config)
+	refStore, err := refstore.NewBoltRefStore(config.DatabasePath)
+
+	if err != nil {
+		log.Error(err)
+	}
+
+	handler := handler.NewGitlabHookHandler(config, commitLogger, notifier, refStore)
+
+	srv.AddRoute("/hooks/commit-message", handler)
```
[the remark this answers is not in the document]
Comment (fetzi) [reply internal/app/commitbot/commitbot.go]: I'm not sure if all hooks should be available under the same url? maybe `/hooks/gitlab/commit-message`?

File: internal/pkg/handler/handler.go (hunk 2 above)
Comment (fetzi) [reply internal/pkg/handler/handler.go]: I've removed the ref value completely

File: conf/sample.toml [hunk 10]
```diff
@@ -0,0 +1,24 @@
+[Server]
+LogLevel = "debug"
```
Comment (Munsio) [inline conf/sample.toml]: Should not belong to the server setting

File: conf/sample.toml [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+Token = "SECRET-TOKEN"
+
+[CommitLog]
+Type = "logstash" # Type can also be set to "file", this will output the commit log info on stdout
+Server = "LOGSTASH-SERVER"
+Servicename = "commit-message"
+
+[Slack]
+Token = "SLACK-OAUTH-TOKEN"
+Emoji = ":cop:"
```
Comment (Munsio) [inline conf/sample.toml]: Should be called `IconEmoji` - it is called like that on the official slack documentation so people would know what it is. Emoji can be something that be append on the message etc.

File: internal/app/commitbot/commitbot.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Munsio) [reply internal/app/commitbot/commitbot.go]: As discussed before - i would like to change it to /hooks/gitlab cause those bots are running on different ports

**Pull request merged**